test/unittests/state_transition.cpp

```diff
@@ -99,46 +95,64 @@ void state_transition::TearDown()
         EXPECT_EQ(trace_capture->get_capture(), expect.trace);
     }
 
+    TestState post{state};
+
     for (const auto& [addr, expected_acc] : expect.post)
     {
-        const auto acc = state.find(addr);
-        if (!expected_acc.exists)
+        const auto ait = post.find(addr);
+        if (ait == post.end())
         {
-            EXPECT_EQ(acc, nullptr) << "account " << addr << " should not exist";
+            EXPECT_FALSE(expected_acc.exists) << addr << ": should exist";
+            continue;
         }
-        else
+
+        ASSERT_TRUE(expected_acc.exists) << addr << ": should not exist";
+        const auto& acc = ait->second;
+
+        if (expected_acc.nonce.has_value())
         {
-            ASSERT_NE(acc, nullptr) << "account " << addr << " should exist";
-            if (expected_acc.nonce.has_value())
-            {
-                EXPECT_EQ(acc->nonce, *expected_acc.nonce) << "account " << addr;
-            }
-            if (expected_acc.balance.has_value())
-            {
-                EXPECT_EQ(acc->balance, *expected_acc.balance)
-                    << to_string(acc->balance) << " vs " << to_string(*expected_acc.balance)
-                    << " account " << addr;
-            }
-            if (expected_acc.code.has_value())
-            {
-                EXPECT_EQ(acc->code, *expected_acc.code) << "account " << addr;
-            }
-            for (const auto& [key, value] : expected_acc.storage)
+            EXPECT_EQ(acc.nonce, *expected_acc.nonce) << addr;
+        }
+        if (expected_acc.balance.has_value())
+        {
+            EXPECT_EQ(acc.balance, *expected_acc.balance)
+                << addr << ": balance " << to_string(acc.balance) << " vs "
+                << to_string(*expected_acc.balance);
+        }
+        if (expected_acc.code.has_value())
+        {
+            EXPECT_EQ(acc.code, *expected_acc.code) << addr << ": wrong code";
+        }
+        for (const auto& [key, value] : expected_acc.storage)
+        {
+            auto sit = acc.storage.find(key);
+            if (value)
             {
-                EXPECT_EQ(acc->storage[key].current, value) << "account " << addr << " key " << key;
+                ASSERT_NE(sit, acc.storage.end()) << addr << ": missing key " << key;
+                EXPECT_EQ(sit->second, value) << addr << ": key " << key;
             }
-            for (const auto& [key, value] : acc->storage)
+            else
             {
-                // Find unexpected storage keys. This will also report entries with value 0.
-                EXPECT_TRUE(expected_acc.storage.contains(key))
-                    << "unexpected storage key " << key << "=" << value.current << " in " << addr;
+                // FIXME:
```

Comment:
is this fixme ok to be left behind? I'm staring at this and can't seem to figure it out, pls check

Reply:
No. I need to figure it out. This code has been changed since the start of this PR. The new types makes the comparison easier (e.g. `TestState{} == TestState{}`) but we also want to have rich failure comments.

Reply:
Fixed. Mostly in #920.